### PR TITLE
[FLINK-8560] Access to the current key in ProcessFunction#OnTimerContext after keyBy()

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

Currently, it is required to store the key of a keyBy() in the processElement method to have access to it in the OnTimerContext.

This is not so good as you have to check in the processElement method for every element if the key is already stored and set it if it's not already set.

A possible solution would adding OnTimerContext#getCurrentKey() or a similar method. Maybe having it in the open() method could maybe work as well.

## Brief change log

added `OnTimerContext#getCurrentKey()`

One limitation is that this impl of `getCurrentKey()` currently is not strongly typed. Declaring the key's type requires adding a new generic type to `ProcessFunction` - making the declaration from `ProcessFunction<IN, OUT>` to `ProcessFunction<IN, K, OUT>`. I'm worried it may break user's application, so I decide to make `getCurrentKey()` return an object. I'd like to discuss the feasibility of having strong type.

## Verifying this change

This change is already covered by existing tests, such as *KeyedProcessOperatorTest*.

## Does this pull request potentially affect one of the following parts:

none

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs / JavaDocs)

### CODE REVIEW

**Comment by bowenli86** [comment]
cc @pnowojski

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java [hunk 1]
```diff
@@ -397,17 +396,16 @@ public void processElement(Integer value, Context ctx, Collector<String> out) th
...
 
 	private static class TriggeringFlatMapFunction extends ProcessFunction<Integer, Integer> {
 
 		private static final long serialVersionUID = 1L;
 
+		static final int TEST_VALUE = 17;
+
 		private final TimeDomain timeDomain;
 
 		public TriggeringFlatMapFunction(TimeDomain timeDomain) {
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java]: rename `timeDomain` -> `expectedTimeDomain` and add `expectedKey` and add assertion for the expected key to `onTimer()` method triggered both in `KeyedProcessOperatorTest#testEventTimeTimers` and `KeyedProcessOperatorTest#testProcessingTimeTimers`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala [hunk 2]
```diff
@@ -162,7 +162,7 @@ class GroupAggProcessFunction(
 
   override def onTimer(
       timestamp: Long,
-      ctx: ProcessFunction[CRow, CRow]#OnTimerContext,
+      ctx: ProcessFunction[CRow, CRow]#OnTimerContext[_],
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala]: I'm not a scala expert, but is this change somehow related to adding the `getCurrentKey()` method?

**Comment by aljoscha** [comment]
@bowenli86 You can take a look at `KeyedBroadcastProcessFunction`. I think we could also add a `KeyedProcessFunction`, that has the key type in the signature. This would allow exposing the key in `onTimer()` and other methods.

**Comment by bowenli86** [comment]
@aljoscha Great idea! I think that's the way to go! Thanks for bringing it up!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala (hunk 2 above)
Comment (bowenli86) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala]: yes, I think I'll create a `KeyedProcessFunction` as Aljoscha suggested, and we probably don't need these scala changes any more.

**Comment by bowenli86** [comment]
@aljoscha @pnowojski  Guys, quick question. I'm about to develop `KeyedProcessFunction` and its operator in a keyed stream. But I found there's already a `KeyedProcessOperator` which is for `ProcessFunction` in a keyed stream. Shall I create a new operator named something like `KeyedProcessFunctionOperator`?

Thanks,
Bowen

**Comment by aljoscha** [comment]
I think you can rename the existing one to `LegacyKeyedProcessOperator` or something like this and have a comment that describes the situation.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java [hunk 3]
```diff
@@ -377,119 +375,114 @@ public T getKey(T value) throws Exception {
...
 
-	private static class TriggeringFlatMapFunction extends ProcessFunction<Integer, Integer> {
+	private static class TriggeringFlatMapFunction extends KeyedProcessFunction<Integer, Integer, Integer> {
 
 		private static final long serialVersionUID = 1L;
 
-		private final TimeDomain timeDomain;
+		private final TimeDomain expectedTimeDomain;
+
+		static final Integer EXPECTED_KEY = 17;
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java]: As in the second PR: does this have to be a static field? Or can we initialise it in the constructor? I think it should work with the `expectedKey` set in the constructor as long as this is not an `ITCase` - and it's not, it's using test harness.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java [hunk 4]
```diff
@@ -377,119 +375,114 @@ public T getKey(T value) throws Exception {
...
 			out.collect("STATE:" + getRuntimeContext().getState(state).value());
 		}
 	}
 
-	private static class BothTriggeringFlatMapFunction extends ProcessFunction<Integer, String> {
+	private static class BothTriggeringFlatMapFunction extends KeyedProcessFunction<Integer, Integer, String> {
 
 		private static final long serialVersionUID = 1L;
 
+		static final Integer EXPECTED_KEY = 5;
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/KeyedProcessOperatorTest.java]: ditto

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java [hunk 5]
```diff
@@ -654,6 +648,63 @@ public ExecutionConfig getExecutionConfig() {
...
+	 * Applies the given {@link KeyedProcessFunction} on the input stream, thereby creating a transformed output stream.
+	 *
+	 * <p>The function will be called for every element in the input streams and can produce zero
+	 * or more output elements.
+	 *
+	 * @param keyedProcessFunction The {@link ProcessFunction} that is called for each element in the stream.
+	 *
+	 * @param <K> The type of key in {@code KeyedProcessFunction}.
+	 *
+	 * @param <R> The type of elements emitted by the {@code PKeyedProcessFunction}.
```
Comment (juergenthomann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java]: Could this be a typo with **P**KeyedProcessFunction instead of KeyedProcessFunction?

**Comment by bowenli86** [comment]
cc @pnowojski  @aljoscha

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 6]
```diff
@@ -686,6 +686,27 @@ class DataStream[T](stream: JavaStream[T]) {
     asScalaStream(javaStream.process(processFunction, implicitly[TypeInformation[R]]))
   }
 
+  /**
+    * Applies the given [[KeyedProcessFunction]] on the input stream, thereby
```
Comment (pnowojski) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala]: ditto

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 7]
```diff
@@ -666,18 +666,18 @@ class DataStream[T](stream: JavaStream[T]) {
...
-   * Applies the given [[ProcessFunction]] on the input stream, thereby
-   * creating a transformed output stream.
-   *
-   * The function will be called for every element in the stream and can produce
-   * zero or more output.
-   *
-   * @param processFunction The [[ProcessFunction]] that is called for each element
-   *                   in the stream.
-   */
+    * Applies the given [[ProcessFunction]] on the input stream, thereby
```
Comment (pnowojski) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala]: Can you revert this formatting? I think proper java docs should be as this was before:
```
/**
 *
 */
```
instead of:
```
/**
  *
  */
```

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 8]
```diff
@@ -79,7 +78,34 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
 
     asScalaStream(javaStream.process(processFunction, implicitly[TypeInformation[R]]))
   }
-  
+
+  /**
+    * Applies the given [[KeyedProcessFunction]] on the input stream, thereby
```
Comment (pnowojski) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: ditto

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala [hunk 9]
```diff
@@ -448,11 +447,35 @@ class DataStreamTest extends AbstractTestBase {
     val flatMapped = src.keyBy(x => x).process(processFunction)
 
     assert(processFunction == getFunctionForDataStream(flatMapped))
+    assert(getOperatorForDataStream(flatMapped).isInstanceOf[LegacyKeyedProcessOperator[_, _, _]])
+  }
+
+  /**
+    * Verify that a [[KeyedStream.process(KeyedProcessFunction)]] call is correctly
```
Comment (pnowojski) [inline flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala]: ditto

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala [hunk 10]
```diff
@@ -473,6 +496,28 @@ class DataStreamTest extends AbstractTestBase {
     assert(getOperatorForDataStream(flatMapped).isInstanceOf[ProcessOperator[_, _]])
   }
 
+  /**
+    * Verify that a [[DataStream.process(KeyedProcessFunction)]] call is correctly
```
Comment (pnowojski) [inline flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala]: ditto

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java [hunk 11]
```diff
@@ -654,6 +648,63 @@ public ExecutionConfig getExecutionConfig() {
...
+	 * @param keyedProcessFunction The {@link ProcessFunction} that is called for each element in the stream.
+	 *
+	 * @param <K> The type of key in {@code KeyedProcessFunction}.
+	 *
+	 * @param <R> The type of elements emitted by the {@code KeyedProcessFunction}.
+	 *
+	 * @return The transformed {@link DataStream}.
+	 */
+	@PublicEvolving
+	public <K, R> SingleOutputStreamOperator<R> process(KeyedProcessFunction<K, T, R> keyedProcessFunction) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java]: Does it make sense to add `process` with a `KeyedProcessFunction` on non keyed `DataStream`?

**Comment by bowenli86** [comment]
Thanks for the review and suggestions. And your comment on `DataStream#process(KeyedProcessFunction)` makes sense, I've removed it. 

(btw, I feel https://github.com/apache/flink/pull/5500 is more urgent that this PR. Can you take it look at that one?)

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 12]
```diff
@@ -54,21 +54,20 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
...
+   * zero or more output. The function can also query the time and set timers. When
+   * reacting to the firing of set timers the function can emit yet more elements.
+   *
+   * The function will be called for every element in the input streams and can produce zero
+   * or more output elements. Contrary to the [[DataStream#flatMap(FlatMapFunction)]]
+   * function, this function can also query the time and set timers. When reacting to the firing
+   * of set timers the function can directly emit elements and/or register yet more timers.
+   *
+   * @param processFunction The [[ProcessFunction]] that is called for each element in the stream.
+   */
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: Revert all reformatings (indent by 1 space) and also add the `@deprecated` annotation with the correct, non-deprecated alternative, as done in the corresponding `java` class.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 13]
```diff
@@ -79,7 +78,34 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
...
+   * The function will be called for every element in the input streams and can produce zero
+   * or more output elements. Contrary to the [[DataStream#flatMap(FlatMapFunction)]]
+   * function, this function can also query the time and set timers. When reacting to the firing
+   * of set timers the function can directly emit elements and/or register yet more timers.
+   *
+   * @param keyedProcessFunction The [[KeyedProcessFunction]] that is called for each element
+   *                             in the stream.
+   */
+  @PublicEvolving
+  def process[K, R: TypeInformation](
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: As in java, you do not need to redefine the `K` here. So you can remove it `def process[R: TypeInformation](`...

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 14]
```diff
@@ -79,7 +78,34 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
...
+    keyedProcessFunction: KeyedProcessFunction[K, T, R]): DataStream[R] = {
+
+    if (keyedProcessFunction == null) {
+      throw new NullPointerException("ProcessFunction must not be null.")
+    }
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: The message now should be `"KeyedProcessFunction must not be null."`

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 15]
```diff
@@ -66,9 +66,9 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
     * function, this function can also query the time and set timers. When reacting to the firing
     * of set timers the function can directly emit elements and/or register yet more timers.
     *
-    * @param processFunction The [[ProcessFunction]] that is called for each element
-    *                   in the stream.
+    * @param processFunction The [[ProcessFunction]] that is called for each element in the stream.
```
Comment (kl0u) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: Please also add that the user now should use the new `KeyedProcessFunction` instead.

**Comment by bowenli86** [comment]
@kl0u I added the comments for `@deprecated` in the javadoc. Let me know if you can merge the two related PRs. Thanks

File: docs/dev/stream/operators/process_function.md [hunk 16]
```diff
@@ -242,4 +242,17 @@ class CountWithTimeoutFunction extends ProcessFunction[(String, String), (String
...
+that are using this incorrect event-time timestamp will fail, and users should adapt their jobs to the correct logic.
+
+## The KeyedProcessFunction
+
+`KeyedProcessFunction`, as an extension of `ProcessFunction`, gives access to the key of timers in its `onTimer(...)`
+method.
+
+{% highlight java %}
+@Override
+public void onTimer(long timestamp, OnTimerContext<K> ctx, Collector<OUT> out) throws Exception {
```
Comment (aljoscha) [inline docs/dev/stream/operators/process_function.md]: I believe this is now `public void onTimer(long timestamp, OnTimerContext<K> ctx, Collector<OUT> out)`, right? @kl0u you could fix this while merging.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/KeyedProcessOperator.java [hunk 17]
```diff
@@ -70,21 +69,15 @@ public void open() throws Exception {
 	@Override
 	public void onEventTime(InternalTimer<K, VoidNamespace> timer) throws Exception {
 		collector.setAbsoluteTimestamp(timer.getTimestamp());
-		onTimerContext.timeDomain = TimeDomain.EVENT_TIME;
-		onTimerContext.timer = timer;
-		userFunction.onTimer(timer.getTimestamp(), onTimerContext, collector);
-		onTimerContext.timeDomain = null;
-		onTimerContext.timer = null;
+		reinitialize(userFunction, TimeDomain.EVENT_TIME, timer);
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/KeyedProcessOperator.java]: Hate to be picky, but I think the name is a bit misleading and we could probably put all of this in a method `invokeUserTime()` that does what `reinitialise()` and `reset()` do.

@kl0u I think you can quickly fix that when merging.

File: docs/dev/stream/operators/process_function.md [hunk 18]
```diff
@@ -242,4 +242,17 @@ class CountWithTimeoutFunction extends ProcessFunction[(String, String), (String
...
-that are using this incorrect event-time timestamp will fail, and users should adapt their jobs to the correct logic.
\ No newline at end of file
+that are using this incorrect event-time timestamp will fail, and users should adapt their jobs to the correct logic.
+
+## The KeyedProcessFunction
+
+`KeyedProcessFunction`, as an extension of `ProcessFunction`, gives access to the key of timers in its `onTimer(...)`
+method.
+
+{% highlight java %}
```
Comment (aljoscha) [inline docs/dev/stream/operators/process_function.md]: Maybe also add Scala example code.

Comment (kl0u) [reply docs/dev/stream/operators/process_function.md]: @bowenli86 As soon as the scala example is added, I can take care of the other two comments and merge! Let me know when you update the PR, and thanks for the work!

**Comment by bowenli86** [comment]
@kl0u @aljoscha   I added the scala example, and I believe the only build failure in Travis is irrelevant

**Comment by kl0u** [comment]
Thanks @bowenli86 ! I will merge as soon as Travis gives green.
